File: ts_src/replace.ts
// Copyright (c) 2010-2020 Sven Kreiss, Kyle Cranmer

import { combiningmarks, replacements, subsuperscripts } from './data';


export function replace(f: string): string {
    // Catch cases like \not\subset and \not\in and convert them to
    // use the combining character slash as in \slash{\subset}
    f = f.replace(/\\not(\\[A-z]+)/g, '\\slash{$1}');
    // escape combining marks with a space after the backslash
    for (const ic in combiningmarks) {
        const c = combiningmarks[ic];

        let i = -1;
        while (
            (i = f.indexOf(c[0], i+1)) > -1
            && f.indexOf("}", i+1) > i
        ) {
            f = f.slice(0, i+1) + ' ' + f.slice(i+1);
        }
    }

    // console.log(replacements);
    for (const ir in replacements) {
        const r = replacements[ir];
        // dirty way of a replaceAll():
        f = f.split(r[0]).join(r[1]);

        if (r[0].slice(-2) == '{}') {
            f = f.split('\\ '+r[0].slice(1)).join(r[1]);
        }
    }

    // expand groups of subscripts: _{01234}
    let isub = -1;
    while (
        (isub = f.indexOf("_{", isub+1)) > -1
        && f.indexOf("}", isub+1) > isub
    ) {
        f = f.slice(0, isub) + '_' + f[isub+2] + '_{' + f.slice(isub+3);
        f = f.replace('_{}', '');
    }

    // expand groups of superscripts: ^{01234}
    let isup = -1;
    while (
        (isup = f.indexOf("^{", isup+1)) > -1
        && f.indexOf("}", isup+1) > isup
    ) {
        f = f.slice(0, isup) + '^' + f[isup+2] + '^{' + f.slice(isup+3);
        f = f.replace('^{}', '');
    }

    // now replace subsuperscripts
    for (const ir in subsuperscripts) {
        const r = subsuperscripts[ir];
        // dirty way of a replaceAll():
        f = f.split(r[0]).join(r[1]);
    }

    // combining marks (unicode char modifies previous char)
    for (const ic in combiningmarks) {
        const c = combiningmarks[ic];

        let i = -1;
        while (
            (i = f.indexOf('\\ '+c[0].slice(1)+'{', i+1)) > -1
            && f.indexOf("}", i+1) > i
        ) {
            const newString = f[i+c[0].length+2] + c[1];
            f = f.slice(0,i)+newString+f.slice(i+1+c[0].length+3);
        }
    }

    return f;
}

File: unicodeit/replace.py
# Copyright (c) 2010 Sven Kreiss, Kyle Cranmer
import re

from .data import REPLACEMENTS, COMBININGMARKS, SUBSUPERSCRIPTS


def replace(f: str):
    # Catch cases like \not\subset and \not\in and convert them to
    # use the combining character slash as in \slash{\subset}
    f = re.sub(r'\\not(\\[A-z]+)', r'\\slash{\1}', f)
    # escape combining marks with a space after the backslash
    for c in COMBININGMARKS:
        f = f.replace(c[0] + '{', '\\ ' + c[0][1:] + '{')

    # replace
    for r in REPLACEMENTS:
        f = f.replace(r[0], r[1])

        # check whether it was escaped for combining marks but has empty braces
        if r[0].endswith('{}'):
            f = f.replace('\\ ' + r[0][1:], r[1])

    # expand groups of subscripts: \_{01234}
    offset = 0
    for s in re.finditer(
            r"_\{[0-9\+-=\(\)<>\-aeoxjhklmnpstiruv"
            r"\u03B2\u03B3\u03C1\u03C6\u03C7\u2212]+\}", f):
        newstring, n = re.subn(
            r"([0-9\+-=\(\)<>\-aeoxjhklmnpstiruv"
            r"\u03B2\u03B3\u03C1\u03C6\u03C7\u2212])",
            r"_\1", s.group(0)[2:-1])
        f = f[:s.start() + offset] + newstring + f[s.end() + offset:]
        offset += n * 2 - (n + 3)

    # expand groups of superscripts: \^{01234}
    offset = 0
    for s in re.finditer(
            r"\^\{[0-9\+-=\(\)<>ABDEGHIJKLMNOPRTUW"
            r"abcdefghijklmnoprstuvwxyz"
            r"\u03B2\u03B3\u03B4\u03C6\u03C7\u222B\u2212]+\}", f):
        newstring, n = re.subn(
            r"([0-9\+-=\(\)<>ABDEGHIJKLMNOPRTUW"
            r"abcdefghijklmnoprstuvwxyz"
            r"\u03B2\u03B3\u03B4\u03C6\u03C7\u222B\u2212])",
            r"^\1", s.group(0)[2:-1])
        f = f[:s.start() + offset] + newstring + f[s.end() + offset:]
        offset += n * 2 - (n + 3)

    # now replace subsuperscripts
    for r in SUBSUPERSCRIPTS:
        f = f.replace(r[0], r[1])

    # process combining marks first
    for c in COMBININGMARKS:
        escaped_latex = f'\\ {c[0][1:]}{{'
        while escaped_latex in f:
            i = f.index(escaped_latex)
            if len(f) <= i + len(escaped_latex):
                # incomplete: unescape and continue
                f = f[:i] + c[0] + '{'
                continue

            combined_char = f[i + len(escaped_latex)]

            remainder = ''
            if len(f) >= i + len(escaped_latex) + 2:
                remainder = f[i + len(escaped_latex) + 2:]

            f = f[:i] + combined_char + c[1] + remainder

    return f
